feat: Enable global expose with nested paths

bahugo opened the pull request:
Closes #2350

I tried to be as explicit as I could, I modified Mapping attribute to executable_relname (relative path but with no extension), and executable_name() is now evaluated based on relname.

ruben-arts commented on src/global/mod.rs:
```diff
@@ -22,33 +26,30 @@ fn find_executables(prefix: &Prefix, prefix_package: &PrefixRecord) -> Vec<PathB
+                .iter()
+                .filter(|relative_path| is_executable(prefix, relative_path))
+                .filter_map(|path| {
+                    path.iter().last().and_then(OsStr::to_str).map(|name| {
+                        Executable::new(strip_executable_extension(name.to_string()), path.clone())
+                    })
```
We have `pixi_utils::executable_from_path` as a function to do the conversion from path the an executable string.

ruben-arts commented on src/global/project/manifest.rs:
```diff
@@ -431,40 +431,52 @@ impl Manifest {
         Self {
             exposed_name,
-            executable_name,
+            executable_relname: Path::new(&executable_relname)
+                .with_extension("")
+                .to_str()
+                .unwrap_or(&executable_relname)
+                .to_string(),
```
with extention doesn't work for `python3.12` as that will create `python3`. 

You can use `pixi_utils::executable_from_path` here aswell.

bahugo replied:
I've just changed new method, I've used strip_executable_extension (because executable_from_path drop nested directory)

ruben-arts commented on src/global/project/manifest.rs:
```diff
@@ -431,40 +431,52 @@ impl Manifest {
 pub struct Mapping {
     exposed_name: ExposedName,
-    executable_name: String,
+    executable_relname: String,
```
What is the reason to change to `relname`? And what does it mean to you?

bahugo replied:
I'm not a huge fan of this name too. It's a relative path but with no extension, maybe you'll have an idea for a better name.
I want to be able to access both values executable_relname and executable_name  that's why I changed this attribute name.

ruben-arts commented on src/global/project/manifest.rs:
```diff
@@ -431,40 +431,52 @@ impl Manifest {
+    }
     pub fn executable_name(&self) -> &str {
-        &self.executable_name
+        if let Some(executable_file_name) = Path::new(&self.executable_relname).file_name() {
+            return executable_file_name
+                .to_str()
+                .unwrap_or(&self.executable_relname);
+        };
+        &self.executable_relname
```
I believe you can use the `pixi_utils::executable_from_path` again

bahugo replied:
I could have use it but it returns a String not a &str, it would break many things.

bahugo replied:
Here I'm not trimming the extension, self.executable_relname is already trimmed at instantiation (in fn new) and file_name() will not trim python3.12
I can't use .as_str() because it would be a reference to a temporary value created inside the method that won't persist.
If you want I could change executable_from_path and strip_executable_extension to take and return &str but It might be a bit overkill.
What do you prefer?

ruben-arts replied:
I'm fine with it, I cleaned it up a little more in my diff.

ruben-arts reviewed (changes requested):
Thanks for the work! I've left some comments but the general implementation looks good. 

- Could you add an integration test to `tests/integration/test_global`? 
- Could you add this case to the documentation in https://pixi.sh/latest/features/global_tools/#exposed `docs/features/global_tools.md`

If you need help creating a dummy package let us know! It should be pretty straight forward to create your situation in a dummy package and test its logic in there.

bahugo commented:
@ruben-arts I tried to add an integration test.
I modified dummy-channel-1, to a add a nested executable.
I'm not sure about which directory is $PREFIX/bin is it the bin directory were pixi will store exposed executables?
Anyway, I added rattler-build dependency to pytest feature and ran `pixi r update-integration-test-data` but it failed:
```
│ ╭─ Running build script
 │ │ %SRC_DIR%>IF "" == "" (
 │ │
 │ │  call %SRC_DIR%\build_env.bat
 │ │ )
 │ │ La syntaxe de la commande n’est pas correcte.
 │ │ %SRC_DIR%>mkdir -p $PREFIX/bin
 │ │ %SRC_DIR%>echo "dummy_e on windows"  1>$PREFIX/bin/dummy_e.bat
 │ │ Le chemin d’accès spécifié est introuvable.
 │ │ × error Script failed with status 1
 │ │ × error Work directory: 'C:\Users\bahugo\dev\pixi\tests\integration\test_data\channels\dummy_channel_1\bld\rattler-build_dummy_e_1730147400\work'
 │ │ × error To debug the build, run it manually in the work directory (execute the `./conda_build.sh` or `conda_build.bat` script)
 │ │
 │ ╰─────────────────── (took 0 seconds)
 │
 ╰─────────────────── (took 0 seconds)
Error:   × Script failed

Traceback (most recent call last):
  File "C:\Users\bahugo\dev\pixi\tests\integration\test_data\update-channels.py", line 37, in <module>
    main()
  File "C:\Users\bahugo\dev\pixi\tests\integration\test_data\update-channels.py", line 18, in main
    subprocess.run(
  File "C:\Users\bahugo\dev\pixi\.pixi\envs\default\Lib\subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['rattler-build', 'build', '--target-platform', 'win-64', '--output-dir', 'channels/dummy_channel_1', '--recipe', 'recipes/dummy_channel_1.yaml']' returned non-zero exit status 1.
```
I tried to run conda_build.bat but it failed too.
I'm not familiar with rattler-build, do you have an idea?

bahugo commented:
> @ruben-arts I tried to add an integration test. I modified dummy-channel-1, to a add a nested executable. I'm not sure about which directory is $PREFIX/bin is it the bin directory were pixi will store exposed executables? Anyway, I added rattler-build dependency to pytest feature and ran `pixi r update-integration-test-data` but it failed:
> 
> ```
> │ ╭─ Running build script
>  │ │ %SRC_DIR%>IF "" == "" (
>  │ │
>  │ │  call %SRC_DIR%\build_env.bat
>  │ │ )
>  │ │ La syntaxe de la commande n’est pas correcte.
>  │ │ %SRC_DIR%>mkdir -p $PREFIX/bin
>  │ │ %SRC_DIR%>echo "dummy_e on windows"  1>$PREFIX/bin/dummy_e.bat
>  │ │ Le chemin d’accès spécifié est introuvable.
>  │ │ × error Script failed with status 1
>  │ │ × error Work directory: 'C:\Users\bahugo\dev\pixi\tests\integration\test_data\channels\dummy_channel_1\bld\rattler-build_dummy_e_1730147400\work'
>  │ │ × error To debug the build, run it manually in the work directory (execute the `./conda_build.sh` or `conda_build.bat` script)
>  │ │
>  │ ╰─────────────────── (took 0 seconds)
>  │
>  ╰─────────────────── (took 0 seconds)
> Error:   × Script failed
> 
> Traceback (most recent call last):
>   File "C:\Users\bahugo\dev\pixi\tests\integration\test_data\update-channels.py", line 37, in <module>
>     main()
>   File "C:\Users\bahugo\dev\pixi\tests\integration\test_data\update-channels.py", line 18, in main
>     subprocess.run(
>   File "C:\Users\bahugo\dev\pixi\.pixi\envs\default\Lib\subprocess.py", line 571, in run
>     raise CalledProcessError(retcode, process.args,
> subprocess.CalledProcessError: Command '['rattler-build', 'build', '--target-platform', 'win-64', '--output-dir', 'channels/dummy_channel_1', '--recipe', 'recipes/dummy_channel_1.yaml']' returned non-zero exit status 1.
> ```
> 
> I tried to run conda_build.bat but it failed too. I'm not familiar with rattler-build, do you have an idea?

@ruben-arts I've managed to update integration test channel by compiling them use debian in wsl.

I ran integration tests on windows, the test I added ran successfully, but I got a fail for another test:

```
E           AssertionError: Return code was 106, expected 0, stderr: No pyvenv.cfg file

tests\integration\common.py:55: AssertionError
```
I don't get why it failed.

ruben-arts commented on docs/features/global_tools.md:
````diff
@@ -113,6 +113,25 @@ exposed = { ansible = "ansible" } # (1)!
+```
+you can also omit the extension
+```
+pixi global install dotnet --expose dotnet=dotnet\dotnet
````
I tried this example on Linux but that package doesn't seem to work. Do you understand why the `dotnet` setup is using a non `bin/xx` setup for it's binaries?

bahugo replied:
I have no Idea.
I had never tried to install dotnet on linux but your right it's installed in `.pixi/envs/dotnet/lib/dotnet/dotnet`
For some reason everything is painful with windows tools...
Maybe a more cross platform or generic tool would be better for the documentation.

ruben-arts replied:
Yeah, I contacted @dhirschfeld as he is a pixi user and the maintainer of `dotnet` maybe he has some smart things to say about this.

dhirschfeld replied:
Not sure about smart - the TL;DR is that it's just a binary repackage of upstream and that seems to be the way it's packaged by Microsoft.

The package sets a number of env vars in an activation script, including the PATH. I'd like to port setting/unsetting the env vars to the JSON format (assuming that works with `pixi`?) but haven't gotten around to it.

ruben-arts replied:
Yes pixi supports that :+1:

ruben-arts replied:
Is there another example package, where this would be required, that does work on linux?

ruben-arts commented:
While reviewing I checked it out and did some small changes, here is the diff patch you could apply:

ruben-arts commented:
ff

bahugo commented:
Something is not working with python activation in integration test, I don't know why...
```bat
  Windows 10.0.26100 on D049 on  main +/- is 📦 v0.34.0 🦀 v1.81.0
~\dev\pixi ➜ C:\Users\bahugo\dev\pixi\pytest-temp\popen-gw0\test_sync_dependencies0\bin\python-injected.bat --version
> No pyvenv.cfg file

  Windows 10.0.26100 on D049 on  main +/- is 📦 v0.34.0 🦀 v1.81.0
~\dev\pixi ❯ C:\Users\bahugo\dev\pixi\pytest-temp\popen-gw0\test_sync_dependencies0\envs\test\Lib\venv\scripts\nt\python.exe
> No pyvenv.cfg file
```

ruben-arts commented:
I'll give this a go today! See if I can fix the issues

ruben-arts commented:
I'm sorry you are experiencing such a crazy merge conflict train. The code works well, I hope you have another example we can use in the documentation.

ruben-arts commented:
If fixed the merge issue, and tested some more.

Your implementation allows to use the following command on both platforms:
```
pixi global install pyside6 --expose qmlls=qt6/bin/qmlls
```

However it doesn't run anything on my machine. I'm fine with skipping an useful example as I'm assuming its not a big set of users hitting this. And other wise we'll find a example through that.

Hofer-Julian reviewed (approved):
Thanks, @bahugo and @ruben-arts, great work!

ruben-arts commented:
There is still an issue with this logic, the `pixi global install python` now exposes a  different version of `python.exe` as there are two in the whole binary. We need to get rid of the nested logic if the expose name is not a path like object.

bahugo commented:
> There is still an issue with this logic, the `pixi global install python` now exposes a different version of `python.exe` as there are two in the whole binary. We need to get rid of the nested logic if the expose name is not a path like object.

@ruben-arts I don't think that nested path is the reason why it fails for python on windows.
In global/mod.rs I've changed executable search to find every executables like in every folder to be able to expose them if needed.
I think I must add a filter for cases where 2 executables where found, in those cases I must keep the one located in one of those folders : 
```rust
let binary_folders = if cfg!(windows) {
        &([
            "",
            "Library/mingw-w64/bin/",
            "Library/usr/bin/",
            "Library/bin/",
            "Scripts/",
            "bin/",
        ][..])
    } else {
        &(["bin"][..])
    };
```
I'll give it a try tomorrow.

bahugo commented:
@ruben-arts @Hofer-Julian it's finally working!

Hofer-Julian reviewed (approved):
Thanks @bahugo!
